source/http_client_connection.c

```diff
@@ -0,0 +1,551 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include "http_client_connection.h"
+
+#include "io.h"
+
+#include <aws/common/array_list.h>
+#include <aws/http/request_response.h>
+#include <aws/io/socket.h>
+
+#include <memoryobject.h>
+
+const char *s_capsule_name_http_client_connection = "aws_http_client_connection";
+const char *s_capsule_name_http_client_stream = "aws_http_client_stream";
+
+struct py_http_connection {
+    struct aws_allocator *allocator;
+    struct aws_http_connection *connection;
+    PyObject *capsule;
+    PyObject *on_connection_setup;
+    PyObject *on_connection_shutdown;
+    bool destructor_called;
+    bool shutdown_called;
+};
+
+static void s_http_client_connection_destructor(PyObject *http_connection_capsule) {
+    struct py_http_connection *http_connection =
+        PyCapsule_GetPointer(http_connection_capsule, s_capsule_name_http_client_connection);
+    assert(http_connection);
+    http_connection->destructor_called = true;
+    if (http_connection->connection) {
+
+        if (aws_http_connection_is_open(http_connection->connection)) {
+            aws_http_connection_close(http_connection->connection);
+        }
+
+        aws_http_connection_release(http_connection->connection);
+        http_connection->connection = NULL;
+    }
+
+    if (http_connection->shutdown_called) {
+        aws_mem_release(http_connection->allocator, http_connection);
+    }
+}
+
+static void s_on_client_connection_setup(struct aws_http_connection *connection, int error_code, void *user_data) {
+
+    struct py_http_connection *py_connection = user_data;
+    PyGILState_STATE state = PyGILState_Ensure();
+    PyObject *result = NULL;
+    PyObject *capsule = NULL;
+
+    if (!error_code) {
+        py_connection->connection = connection;
+        capsule =
+            PyCapsule_New(py_connection, s_capsule_name_http_client_connection, s_http_client_connection_destructor);
+        py_connection->capsule = capsule;
+    } else {
+        aws_mem_release(py_connection->allocator, py_connection);
+    }
+
+    result = PyObject_CallFunction(py_connection->on_connection_setup, "(Ni)", capsule, error_code);
```

Comment:
If there was an error, you just released the py_connection above, so this is going to fail. Maybe store on_connection_setup into a temporary before deletion?